src/v/security/gssapi_principal_mapper.cc

```diff
@@ -283,6 +359,42 @@ ss::sstring gssapi_rule::replace_substitution(
     return replace_value;
 }
 
+gssapi_principal_mapper::gssapi_principal_mapper(
+  std::string_view default_realm,
+  config::binding<std::optional<std::vector<ss::sstring>>>
+    principal_to_local_rules_cb)
+  : _default_realm(default_realm)
+  , _principal_to_local_rules_binding(std::move(principal_to_local_rules_cb))
+  , _rules{detail::parse_rules(
+      _default_realm, _principal_to_local_rules_binding())} {
+    _principal_to_local_rules_binding.watch([this]() {
+        _rules = detail::parse_rules(
+          _default_realm, _principal_to_local_rules_binding());
+    });
+}
+
+std::optional<ss::sstring>
+gssapi_principal_mapper::apply(gssapi_name name) const {
+    std::vector<std::string_view> params;
+    if (name.host_name().empty()) {
+        if (name.realm().empty()) {
+            return name.service_name();
+        }
+        params = {name.realm(), name.service_name()};
+    } else {
+        params = {name.realm(), name.service_name(), name.host_name()};
+    }
+
+    for (const auto& r : _rules) {
+        if (auto result = r.apply(params); result.has_value()) {
+            return result;
+        }
+    }
+
+    vlog(seclog.error, "No rules apply to {}, rules: {}", name, _rules);
```

Comment:
is this really an error? the message doesn't sound like an error, but i'm also not certain of what the scenario is.

Reply:
That's fair.  I feel like it's "more important" than an informational log so maybe warning level?

Reply:
maybe. i sent you a link to log level guide on slack

Reply:
Pushed updates and commented (in the review) the `vlogs` with my justification for the levels.  Happy to hear your take on it.